frontend/hub/api/utils.tsx

```diff
@@ -7,11 +7,49 @@ import {
   postHubRequest,
   putHubRequest,
 } from './request';
+import { useTranslation } from 'react-i18next';
+import { requestGet } from '../../common/crud/Data';
 
 function getBaseAPIPath() {
   return process.env.HUB_API_PREFIX;
 }
 
+export function useRepositoryBasePath(name: string, pulp_href?: string): Promise<unknown> {
+  const { t } = useTranslation();
+  return Promise.all([
+    pulp_href
+      ? Promise.resolve({ name, pulp_href })
+      : requestGet<{ data: { results: unknown } }>(
+          pulpAPI`/repositories/ansible/ansible/?name=${name}&limit=1`
+        ).then(firstResult),
+    requestGet<{ data: { results: unknown } }>(
+      pulpAPI`/distributions/ansible/ansible/?name=${name}&limit=1`
+    ).then(firstResult),
+  ]).then(async ([repository, distribution]) => {
```

Comment:
@ShaiahWren I took a look at the code and a possible change would something like

```

export async function useRepositoryBasePath(name: string, pulp_href?: string): Promise<unknown> {
  const { t } = useTranslation();

  try {
    let repository;
    if (pulp_href) {
      repository = { name, pulp_href };
    } else {
      const repositoryResponse = await requestGet<{ data: { results: unknown } }>(
        pulpAPI`/repositories/ansible/ansible/?name=${name}&limit=1`
      );
      repository = firstResult(repositoryResponse);
    }
    if (!repository) {
      throw new Error(t`Failed to find repository ${name}`);
    }
    const distributionResponse = await requestGet<{ data: { results: unknown } }>(
      pulpAPI`/distributions/ansible/ansible/?name=${name}&limit=1`
    );
    let distribution = firstResult(distributionResponse);

    if (distribution && distribution.repository === repository.pulp_href) {
      return distribution.base_path;
    }

    const newDistResponse = await requestGet<{ data: { results: unknown } }>(
      pulpAPI`/distributions/ansible/ansible/?repository=${repository.pulp_href}&ordering=pulp_created&limit=1`
    );
    distribution = firstResult(newDistResponse);

    if (!distribution) {
      throw new Error(t`Failed to find a distribution for repository ${name}`);
    }

    return distribution.base_path;
  } catch (error) {
    if (isRequestError(error)) {
      throw error;
    } else {
      throw new Error(t`Failed to find repository base path`);
    }
  }
}


```

It will be required to verify that we are achieving the same with this code. Just an idea how we could refactor this one.

Reply:
This will take twice as long to wait on the server, won't it?

And any exception will cause "Failed to find repository base path", none of the error messages in the if statements will work since those are now exceptions?

Reply:
`await` is not great when working with multiple promises,
and it makes it hard to re-throw the exception on failure.

If you want a more readable version with the same semantics, what about

```js
type Results = { data: { results: Repository[] | Distribution[] } };

export function useRepositoryBasePath(name: string, pulp_href?: string): Promise<string> {
  const { t } = useTranslation();

  const repoRequest = pulp_href ? Promise.resolve({ name, pulp_href }) : requestGet<Results>(
    pulpAPI`/repositories/ansible/ansible/?name=${name}&limit=1`
  ).then(firstResult);

  const distroRequest = requestGet<Results>(
    pulpAPI`/distributions/ansible/ansible/?name=${name}&limit=1`
  ).then(firstResult);

  return Promise.all(repoRequest, distroRequest).then(async ([ repository, distribution ]) => {
    if (!repository) {
      throw new Error(t`Failed to find repository ${name}`);
    }

    if (distribution && distribution.repository === repository.pulp_href) {
      return distribution.base_path;
    }

    distribution = await requestGet<Results>(
      pulpAPI`/distributions/ansible/ansible/?repository=${repository.pulp_href}&ordering=pulp_created&limit=1`
    ).then(firstResult);

    if (!distribution) {
      throw new Error(t`Failed to find a distribution for repository ${name}`);
    }

    return distribution.base_path;
  });
}

function firstResult({ data: { results } }: Results) {
  return results[0];
}

Reply:
Yeah, if we want to keep the requests in parallel we could add back the Promise.all. Not sure how much performative will be that, but if we can make those requests in parallel, +1.

The requests client will raise HTTP RequestError if there is any error underline on the fetch itself.

I like your idea @himdelI it does not need to be part of this PR, but we could have small functions to fetch some resources that we will need to fetch over and over. Something like: 


```js
 const getDistribution = async (distroName: string) => {
    const response = await requestGet<Results>(
      pulpAPI`/distributions/ansible/ansible/?name=${distroName}&limit=1`
    );
    return firstResult(response);
  };

```

Thanks @ShaiahWren and @himdel for looking into improve readability here.